src/ExtendedXmlSerializer/Cache/PropertieDefinition.cs

```diff
@@ -29,24 +29,37 @@ internal class PropertieDefinition
         public PropertieDefinition(Type type, PropertyInfo propertyInfo, string name)
         {
             Name = string.IsNullOrEmpty(name) ? propertyInfo.Name : name;
+            DeclaringType = propertyInfo.DeclaringType;
             TypeDefinition = TypeDefinitionCache.GetDefinition(propertyInfo.PropertyType);
             _getter = ObjectAccessors.CreatePropertyGetter(type, propertyInfo.Name);
-            _propertySetter = ObjectAccessors.CreatePropertySetter(type, propertyInfo.Name);
+            _propertySetter = CreateSetter( type, propertyInfo.Name, propertyInfo.CanWrite );
         }
 
         public PropertieDefinition(Type type, FieldInfo fieldInfo, string name)
         {
             Name = string.IsNullOrEmpty(name) ? fieldInfo.Name : name;
             TypeDefinition = TypeDefinitionCache.GetDefinition(fieldInfo.FieldType);
             _getter = ObjectAccessors.CreatePropertyGetter(type, fieldInfo.Name);
-            _propertySetter = ObjectAccessors.CreatePropertySetter(type, fieldInfo.Name);
+            _propertySetter = CreateSetter( type, fieldInfo.Name, !fieldInfo.IsInitOnly );
+        }
+
+        ObjectAccessors.PropertySetter CreateSetter( Type type, string name, bool writable )
+        {
+            if ( writable )
+            {
+                return ObjectAccessors.CreatePropertySetter(type, name);
+            }
+
+            var result = TypeDefinition.MethodAddToCollection != null ? new ObjectAccessors.PropertySetter( TypeDefinition.MethodAddToCollection ) : null;
```

Comment:
Oh really?  Are you sure you got all the commits?  That should be what adds items to collections in the case of read-only collection properties.